src/python/pants/engine/internals/graph.py

```diff
@@ -101,8 +137,40 @@ async def resolve_target(
 
 @rule
 async def resolve_targets(addresses: Addresses) -> Targets:
```

Comment:
You could save some redundancy by having this rule take `UnexpandedTargets` as input. Saves lines 144 and 145, and simplifies the rule graph.